src/System.Management.Automation/namespaces/FileSystemProvider.cs

```diff
@@ -40,41 +40,49 @@ internal class InodeTracker
         private Dictionary<(UInt64, UInt64), bool>  _visitations;
 
         /// <summary>
-        /// Construct a new InodeTracker
+        /// Construct a new InodeTracker with an initial path
         /// </summary>
-        internal InodeTracker()
+        internal InodeTracker(string path)
         {
             _visitations = new Dictionary<(UInt64, UInt64), bool>();
-        }
-
-        /// <summary>
-        /// Determine if a path has been visited.
-        /// </summary>
-        /// <param name="path">Path to the file or directory to be checked.</param>
-        /// <returns>True if the path has been visited, false otherwise</returns>
-        internal bool IsPathVisited(string path)
-        {
             var inodeData = (0UL, 0UL);
 
             if (InternalSymbolicLinkLinkCodeMethods.GetInodeData(path, out inodeData))
             {
-                return _visitations.ContainsKey(inodeData);
+                _visitations[inodeData] = true;
             }
-            return false;
         }
 
         /// <summary>
-        /// Mark a path as visited.
+        /// Try to mark a path as having been visited.
         /// </summary>
         /// <param name="path">Path to the file or directory to be marked as visited.</param>
-        internal void VisitPath(string path)
+        /// <returns>
+        /// False if the path has already been visited.
+        /// True if the path has not yet been visited.
+        /// </returns>
+        /// <remarks>
+        /// If the path has not yet been visited then before this method returns it marks
+        /// the path as having been visited.
+        /// </remarks>
+        internal bool TryVisitPath(string path)
         {
+            bool rv = true;
             var inodeData = (0UL, 0UL);
 
             if (InternalSymbolicLinkLinkCodeMethods.GetInodeData(path, out inodeData))
             {
-                _visitations[inodeData] = true;
+                if (_visitations.ContainsKey(inodeData))
```

Comment:
Now CoreFX has a new method [Dictionary.TryAdd](https://github.com/dotnet/corefx/blob/3d4d8038d1a3f0c56d3046bc7494f30ecc35e67b/src/Common/tests/System/Collections/DictionaryExtensions.cs) - Could we use them?